File: src/services/service.ts
/**
 * Service control for managing skhd service
 *
 * This service provides:
 * - Getting service status
 * - Starting, restarting, and reloading the service
 * - Managing service state
 */

import { invoke } from '@tauri-apps/api/core';
import type { ServiceStatus } from '../types';

/**
 * Get the current status of the skhd service
 *
 * @returns {Promise<ServiceStatus>} Current service status
 *
 * @example
 * ```typescript
 * const status = await getServiceStatus();
 * console.log('Service state:', status.state);
 * console.log('Service PID:', status.pid);
 * ```
 */
export async function getServiceStatus(): Promise<ServiceStatus> {
  return invoke('get_service_status');
}

export async function startService(): Promise<void> {
  return invoke('start_service');
}

export async function restartService(): Promise<void> {
  return invoke('restart_service');
}

/**
 * Reload the skhd service
 *
 * This will stop and start the service with the current configuration.
 *
 * @throws {string} If reload fails
 *
 * @example
 * ```typescript
 * try {
 *   await reloadService();
 *   console.log('Service reloaded successfully');
 * } catch (error) {
 *   console.error('Failed to reload service:', error);
 * }
 * ```
 */
export async function reloadService(): Promise<void> {
  return invoke('reload_service');
}

File: src-tauri/src/commands/service.rs
use crate::models::ServiceStatus;
use crate::services::ServiceManager;
use tauri::State;

const ACCESSIBILITY_SETTINGS_URL: &str =
    "x-apple.systempreferences:com.apple.preference.security?Privacy_Accessibility";
const INPUT_MONITORING_SETTINGS_URL: &str =
    "x-apple.systempreferences:com.apple.preference.security?Privacy_ListenEvent";

/// Get the current status of the skhd service
#[tauri::command]
pub async fn get_service_status(
    service_manager: State<'_, ServiceManager>,
) -> Result<ServiceStatus, String> {
    service_manager.get_status().await
}

/// Start the skhd service
#[tauri::command]
pub async fn start_service(service_manager: State<'_, ServiceManager>) -> Result<(), String> {
    service_manager.start_service().await
}

/// Stop the skhd service
#[tauri::command]
pub async fn stop_service(service_manager: State<'_, ServiceManager>) -> Result<(), String> {
    service_manager.stop_service().await
}

/// Restart the skhd service
#[tauri::command]
pub async fn restart_service(service_manager: State<'_, ServiceManager>) -> Result<(), String> {
    service_manager.restart_service().await
}

/// Reload the skhd service configuration
#[tauri::command]
pub async fn reload_service(service_manager: State<'_, ServiceManager>) -> Result<(), String> {
    service_manager.reload_service().await
}

/// Install the skhd service (skhd.zig only)
#[tauri::command]
pub async fn install_service(service_manager: State<'_, ServiceManager>) -> Result<(), String> {
    service_manager.install_service().await
}

/// Uninstall the skhd service (skhd.zig only)
#[tauri::command]
pub async fn uninstall_service(service_manager: State<'_, ServiceManager>) -> Result<(), String> {
    service_manager.uninstall_service().await
}

/// Open the macOS Accessibility pane without exposing arbitrary URL opening.
#[tauri::command]
pub fn open_accessibility_settings() -> Result<(), String> {
    tauri_plugin_opener::open_url(ACCESSIBILITY_SETTINGS_URL, None::<&str>)
        .map_err(|error| error.to_string())
}

/// Open the macOS Input Monitoring pane without exposing arbitrary URL opening.
#[tauri::command]
pub fn open_input_monitoring_settings() -> Result<(), String> {
    tauri_plugin_opener::open_url(INPUT_MONITORING_SETTINGS_URL, None::<&str>)
        .map_err(|error| error.to_string())
}
